RFC-0001-Agile Coretime.md

```diff
@@ -0,0 +1,291 @@
+
+All Bulk Coretime sold by Polkadot is done so on a new system parachain known as the *Broker-chain*. Owners of Bulk Coretime are tracked on this chain and the ownership status and properties (i.e. the specific period) of the owned Coretime are exposed over XCM as a non-fungible asset.
+
+At the request of the owner, the Broker-chain allows Bulk Coretime to be:
+
+1. Transferred in ownership.
+2. Split into quantized, non-overlapping segments of Bulk Coretime with the same ownership.
+3. Consumed in exchange for the allocation of a Core during its period.
+4. Consumed in exchange for a pro-rata amount of the revenue from Instantaneous Core sales over its period.
+
+Pre-existing leases SHALL be recorded in the Broker-chain and cores reserved for them.
+
+Sales of Instantaneous Coretime is expected to happen on the Polkadot Relay-chain. The Relay-chain is expected to be responsible for:
+
+- holding non-transferable, non-refundable DOT balance information for collators.
+- setting and adjusting the price of Instantaneous Coretime based on usage.
+- allowing collators to consume their DOT balance in exchange for the ability to schedule one PoV for near-immediate usage.
+- ensuring the Broker-chain has timely book-keeping information on Coretime sales revenue. This should include a total instantaneous revenue amount for each block number.
+
+### Broker-chain
+
+The Broker-chain is a new system parachain. It has the responsibility of providing the Relay-chain via UMP with scheduling information of:
+
+- How many cores which should be made available during the next session.
+- Which cores should be allocated to which para IDs.
+
+Any cores left unallocated are assumed by the Broker-chain to be used for Instantaneous Coretime sales.
+
+It is also expected to receive information from the Relay-chain on total revenue from instantaneous coretime sales on a per-block basis.
+
+### Bulk Sales
+
+There is a periodic sale of Coretime happening at a period of `BULK_PERIOD`.  A number of instances of Coretime spanning from `LEADIN_PERIOD` in the future to `LEADIN_PERIOD + BULK_PERIOD` in the future of a single Polkadot UC Core is offered by the Polkadot System at a fixed price.
+
+These instances which are owned by purchaser are called *regions*. This sale happens on the Broker-chain. Regions are quantized into atomic periods of `TIMESLICE`, into which `BULK_PERIOD` divides a whole number of times. The `Timeslice` type is a `u16` which can be multiplied by `TIMESLICE` to give a `BlockNumber` value indicating the same period in times of (Relay-chain) blocks.
+
+The Broker-chain aims to sell some `BULK_TARGET` of Cores, up to some `BULK_LIMIT`. It makes this sale in a single batch `LEADIN_PERIOD` prior to the beginning of the period being sold. The Broker chain publishes a price for this batch of sales for the `BULK_TARGET` period prior to the sale execution.
```

Comment:
Is there a formula or parameters for how the price adjusts from period to period, based on the actual, limit, and bulk?

I imagine something like a minimum price & an adjustment %. The algorithm should be specified in the RFC to explain how it meets goal (1)

Reply:
The price-change mechanism is possibly out of scope of this RFC. One possibility is for it to have `MAX_PRICE_INCREASE` and `MAX_PRICE_DECREASE` fractions, then a formula:

```
NEW_PRICE := IF CORES_SOLD < BULK_TARGET THEN
    RETURN OLD_PRICE - OLD_PRICE * MAX_PRICE_DECREASE * CORES_SOLD / BULK_TARGET
ELSE
    RETURN OLD_PRICE + OLD_PRICE * MAX_PRICE_INCREASE * (CORES_SOLD - BULK_TARGET) / (BULK_LIMIT - BULK_TARGET)
END IF
```

Reply:
To me that seems quite in-scope for the RFC - as the specifics of the algorithm are likely not configurable by governance, only parameters.

That algorithm looks like it'd work. At least, it does the right thing: price goes up when cores are oversold, and down when cores are undersold. Though we might want some convexity in the adjustment (i.e. price adjusts at a superlinear rate per tick, depending on distance from the target).

Reply:
Indeed - that's why I'd prefer to keep it as a pluggable component which can be proposed and iterated on later.

Reply:
There is now a reference algorithm in the RFC as an illustration of how it could work.